### DISCUSSION TITLE
Are there any existing suites of MJX based manipulation environments with gym style interfaces?

### DISCUSSION THREAD

**Discussion by StoneT2000** [google-deepmind/mujoco#2043] [Asking for Help]
### Intro

Hi!

I am a graduate student at UC San Diego, I am currently investigating and comparing GPU robotics simulators on different manipulation tasks.

### My setup

Ubuntu 22.02, latest MJX/Mujoco python API

### My question

I was wondering if there are any existing manipulation benchmarks built on top of MJX, similar to IsaacLab/ManiSkill 3. I did a quick search on google and here but couldn't find any already built ones. My goal is to run some RL algorithms on the GPU simulation and wanted to try out MJX instead of just the physx based simulators and learn about the differences.

Or does anyone know if e.g. MetaWorld, DM-Control, Gymnasium Robotics have any progress/plans to support MJX? I think I saw gymnasium maintainers discussing using MJX but I have not seen any updates about it since a few months ago.

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by erikfrey** [comment]
Hi @StoneT2000 ,

We spoke with Gymnasium folks a while back - at the time MJX was still missing some features that are now in or will be in shortly.  We'll probably check back in with them now that those are out.

We use MJX for manipulation tasks with the [Aloha platform](https://aloha-2.github.io/) - if it's useful to you to see them, I can get a few of those tasks open sourced.  We also are doing some experiments with [Madrona](https://madrona-engine.github.io/) for vision-based policies, although that is early work that is not yet ready.

Manipulation is an area of active focus for us, so if there's some way we can support your investigation or comparison, please let us know.

**Reply by StoneT2000** [reply]
It would be great if there could some simple mjx manipulation suites open sourced (a cube picking task or precise assembly task would be great). My primary goal at the moment is really for paper writing and comparison between my labs simulator (physx, sapien, maniskill stack) and other sims. While I could maybe write some myself I feel like it could be unfair as I don't know how to tune mjx to be fast while still being realistic (eg for maniskill and isaac lab I tune various sim/rendering hyperparameters to simulate/render as fast as possible without losing too much accuracy)

My lab is also considering whether we should support mjx as a backend in SAPIEN but we honestly still don't understand mjx enough compared to physx GPU sim to make a judgement on it. Hence the need to try out some good manipulation environments.

**Reply by erikfrey** [reply]
I think this is something we can help you with.  Let me talk to a few folks internally and see what we can get out quickly and easily.

Are you familiar with Aloha?  If not, are there any robots in [MuJoCo menagerie](https://github.com/google-deepmind/mujoco_menagerie) that would make your comparison easier?

**Reply by kevinzakka** [reply]
I would really like to see 1-2 ALOHA MJX tasks open-sourced!

**Reply by StoneT2000** [reply]
Thanks @kevinzakka :)

@erikfrey Franka arms work best which I saw in that repo, it is the only one that IsaacSim/Lab, ManiSkill/SAPIEN, and Mujoco simultaneously all support pretty well for manipulation. I don't think there is code for Isaac for any of the other arms (and to my knowledge isaac team tuned the franka arm much more than others, they have 1 arm manipulation task which is using the franka arm which I use for comparison with them).

I have not tried out Aloha but if I can see the Aloha code / setup and whatever optimizations your team might be doing to improve simulation speed / balance accuracy for manipulation, I can copy that over probably to a franka pick cube environment which I have been using as a "standard" benchmarking environment.

**Reply by btaba** [reply]
Hi @StoneT2000 , can you point us to the URDF you're using for ManiSkill/SAPIEN, along with the environment you want to benchmark? We can produce an MJCF in a similar spirit, along with any other optimizations that are required for the current version of MJX to behave in a performant way.

**Reply by StoneT2000** [reply]
@btaba sure thing. It is saved here to panda_v2.urdf and panda_v2.srdf (are srdf files standard? for us we use them to disable some self collisions that never happen to speed up sim, we only disable collision pairs with reason="Default"). Box collisions for grippers to simplify that and then convex meshes for all the other parts. By default adjacent links have collisions disabled implicitly.

https://github.com/haosulab/ManiSkill/blob/main/mani_skill/assets/robots/panda/panda_v2.urdf
https://github.com/haosulab/ManiSkill/blob/main/mani_skill/assets/robots/panda/panda_v2.srdf
meshes: https://github.com/haosulab/ManiSkill/tree/main/mani_skill/assets/robots/panda/franka_description/meshes

That being said, is there a tuned aloha URDF?

**Reply by btaba** [reply]
Hi @StoneT2000 

I took the URDF provided and roughly converted it to [MJCF](https://github.com/btaba/ManiSkill/tree/franka-stonet). A few comments:

*  I removed visual geoms since they are specified via `glb` (does not load in MJ). This shouldn't affect performance on our end since we're not doing rendering in the loop.
* I removed arm self-collisions since convex-convex is not as performant in MJX (something we're actively working on). If I were to continue working on this env, I would add capsules to the arm for self-collisions.
* To speed test using MJX, you can try `python -m mujoco.mjx.testspeed --mjcf=mani_skill/assets/robots/panda/panda_v2_mjx_scene.xml   --base_path=. --batch_size=4096 --nstep=100`

For a tuned Aloha model see https://github.com/google-deepmind/mujoco_menagerie/tree/main/aloha, although you'll see similar modifications as outlined above. Let us know if you have questions!

**Reply by StoneT2000** [reply]
Will this also test manipulation? And the link is the urdf link i sent, where is the mjcf?

**Reply by btaba** [reply]
The link points to a ManiSkill fork, see: https://github.com/btaba/ManiSkill/blob/franka-stonet/mani_skill/assets/robots/panda/panda_v2_mjx_scene.xml

**Reply by StoneT2000** [reply]
Oh I missed that completely, thanks. I'll see if i can benchmark manipulation myself through this then.

**Reply by btaba** [reply]
@StoneT2000 I'm curious what you're looking for in terms of benchmarking manipulation. Is something like the `testspeed` script above sufficient, or are you looking for something else?

**Reply by StoneT2000** [reply]
I think I would like to test the same RL code on mjx and our simulations to try and e.g. pick up a cube (without end to end jit, just torch ppo + mjx env). Alternatively (and probably more accurate), run repeatedly the same trajectory of picking up the cube over and over in parallel and measure that speed/throughput. It doesn't seem like `testspeed` will test other collisions apart from the cube touching the floor? 

Some other questions I have:
- What is the easiest way to turn this mjx scene into a gym style interface?
- How can I define rewards/randomize states? preferrably via python (just for RL benchmarking and comparisons)
- iirc brax previously had a situation where it often would only ever reset once to improve performance, and then during env partial resets it sets state equal to a previously sampled reset state. Is this still the case with mjx and necessary or are full resets possible (e.g. some tasks I build have way too much randomization for just 4096 different reset states to suffice for training with RL).
- What is the easiest way to collect a trajectory with gripper collisions with the cube?

Less important:
- Is there a ready made cartpole env with mjx? 
- I would also like to benchmark rendering although just to double check that is a bit pointless at this moment in time as mjx doesn't have parallel rendering just yet right? (of which then for now I don't plan to include mjx in my rendering benchmarks).

**Reply by btaba** [reply]
Thanks for the explanation! Some responses below:

RE `testspeed`: the collisions are always calculated in MJX for the env I sent you (XLA requires static shapes, and I didn't set them lower than the expected number of contacts with gripper<>cube collisions). I don't think you'll get a big difference in performance with/without contact, but I'd be interested to know what you find.

RE gym style interface: the easiest way is to use [brax](https://github.com/google/brax/blob/main/brax/envs/ant.py), here is an example training Ant with torch and [brax](https://github.com/google/brax/blob/c89a3ad7900648613ead04dedacc91e923e10a68/notebooks/training_torch.ipynb).

RE resets: Yes this is still generally the case in brax, but you can also do "full" resets like [so](https://github.com/google/brax/blob/c89a3ad7900648613ead04dedacc91e923e10a68/brax/training/agents/ppo/train.py#L447-L460). We still haven't seen this making a big difference in practice in our envs. I'm curious what your envs look like.

RE trajectory: you can generate actions via IK to follow a fixed trajectory to pick up the cube. I tested that collisions happen with the gripper/cube, but I didn't tune them s.t. the cube is easily picked up. If you're keen to get a pick-up trajectory with this franka env, I can make sure the cube is easily picked up.

RE cartpole: see the [envs here](https://github.com/google/brax/tree/main/brax/envs), just set the `backend` to 'mjx'

RE rendering: this is very much a WIP, you can find some experimental code [here](https://github.com/shacklettbp/madrona_mjx/blob/aloha/scripts/mjx_aloha_demo.py) .

**Reply by StoneT2000** [reply]
What do you mean always calculated in MJX? If the gripper is not near the cube collisions are still being calculated? So whether there is actual contact or not does not change FPS? I find this surprising as this is not the case with PhysX. Contacts are not computed/considered in the simulation pipeline in gpu sim if two objects are too far away (default is like 2 centimeters in maniskill). Often during RL FPS of the environment goes down as the agent learns to actually grab the intended object more often in e.g. a peg insert task or cube picking task.

My understanding is brax is the RL interfaces for mjx now? What is the difference between the mjx backend and what brax had before?

**Reply by btaba** [reply]
>  If the gripper is not near the cube collisions are still being calculated?

Yes this is currently the case. As mentioned, XLA requires static shapes known at compile time, and the MJCF provided always contains [16](https://github.com/btaba/ManiSkill/blob/86fcc0ce92dce88e7f1d605b357ee2dc8e1e22de/mani_skill/assets/robots/panda/panda_v2_mjx.xml#L8) contact points. We have a [custom parameter](https://github.com/google-deepmind/mujoco/blob/4e3dc14f8ed6a0e50fa7b7fb484d991b7e93a28f/mjx/mujoco/mjx/test_data/shadow_hand/right_hand.xml#L10) to do broadphase, but I didn't add that to the MJCF provided. This is why I mentioned that collisions are always being calculated between the box and the gripper, and that `testspeed` should be a decent representation of the env performance. You can check `mjx_data.contact.geom` on the first step.

> My understanding is brax is the RL interfaces for mjx now? What is the difference between the mjx backend and what brax had before?

Yes we typically use brax as the RL interface for MJX. MJX is MuJoCo implemented in JAX. Other backends in brax implement different physics pipelines (mass-spring, PBD, etc.).

**Reply by StoneT2000** [reply]
Thanks for clarifications! I understand how mjx works much better now, I will try and share benchmarking results when I can.

**Reply by erikfrey** [reply]
I will add one nitpicky addendum: there are scenarios where MJX slows down as more contacts are dynamically encountered during simulation.  The constraint solver can early exit, and will do more work during states where there are more active contacts.  But this slowdown will be less pronounced than PhysX.

**Reply by StoneT2000** [reply]
Yeah I find benchmarking simulation speed incredibly tricky to be honest. It is simple when comparing frameworks with already built tasks (e.g. robosuite vs maniskill) since I would generally assume authors tune their simulations well and cannot be at fault for poor tuning of a different framework. But trying to fairly compare the underlying simulators e.g. mujoco vs sapien vs isaac sim is fairly complicated.

While strange, somehow this makes me think running RL on the environment with the same e.g. PPO code and settings is more or less one of the better ways to compare simulation speeds. E.g. IsaacLab typically uses a low simulation to control frequency ratio = higher raw FPS = slower RL learning as each action moves the robot less distance. Harder to manipulate FPS numbers to an extent if comparing RL wall training time.

Has your team tried to benchmark other simulators before and if so, what is your methodology for fair comparisons?

**Reply by btaba** [reply]
Hi @StoneT2000 

We just open sourced a manipulation env (pickup cube) on the [Franka Panda](https://github.com/google-deepmind/mujoco/commit/0887f54307e93c73510a0812c4ea0ebf148e7838) in the public [MJX tutorial colab](https://colab.sandbox.google.com/github/google-deepmind/mujoco/blob/main/mjx/tutorial.ipynb). I copied some of your reward terms, but not all. I was satisfied enough with the training time so didn't do further tuning. We train for 20M env steps (at a 50Hz control frequency and 200Hz sim frequency). It trains for around 3 minutes on an A100.

Agreed that benchmarking different sims is difficult. In principle I agree about RL being a good benchmark since it tests a real-world workload. However in practice, you'll have to somehow control for differences in [1] the RL algo/impl and [2] reward terms that both may contribute to different sample efficiencies between the set-ups. For example, one could potentially keep tuning the rewards in the Panda env above to get training time down (less env steps), but it's probably not a good use of time since training time is already 3 minutes. As another example, if one were to copy the same exact RL algo/impl and reward terms from a different set-up, I wouldn't be surprised if they still needed to tune RL hypers and maybe rewards due to differences in physics between simulators.

The way I've typically benchmarked between simulators in the past: [1] load the same scene, [2] check that both sims are stable at the same sim and control frequency with random perturbations, [3] ensure contacts occur in the scene for the benchmark time window, and [4] get raw FPS numbers using provided benchmark scripts on the same device. It's not perfect, but it's a bit easier to run through.

**Reply by StoneT2000** [reply]
Thanks I just saw the video on your tweet as well, good to see confirmation and prewritten baseline is working. I'll take a look and play around.

Also agree that RL is not a perfect solution. I think maybe at best what we can do is just share baselines for RL+gpu sim for different frameworks, post raw numbers (that maintainers of respective frameworks can tune themselves as much as they want), share video of results, and let users decide. This is partly a suggestion from the maintainer of [Simulately](https://github.com/geng-haoran/Simulately), which seems to be a bit more of a mostly unbiased repo/wiki for simulators. Perhaps it can be where we just store all benchmarking scripts for all things sim and robotics?

This also ended up being partly what my lab did with our recent maniskill paper release on arxiv. We compared with Isaac Lab and for e.g. rendering fidelity we don't bother to say which simulator's parallel rendering looks better, we just post results and let user decide

**Reply by btaba** [reply]
Thanks for the link to Simulately, happy to provide scripts where they're needed. Also 100% agree, users should (and will) use whatever sim meets their use-case best. We try to advertise [the sharp bits](https://mujoco.readthedocs.io/en/stable/mjx.html#mjx-the-sharp-bits) about MJX as its implemented today. So far MJX has been great for locomotion + sim2real, but not really proven for manipulation workloads. As such, there are bottlenecks in the impl that we're actively working to get rid of!
